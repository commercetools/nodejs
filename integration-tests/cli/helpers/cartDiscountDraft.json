{
  "name": {
    "en": "integration-test-cart-discount"
  },
  "value": {
    "type": "relative",
    "permyriad": 1000
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "target": {
    "type": "lineItems",
    "predicate": "1 = 1"
  },
  "sortOrder": "0.1",
  "requiresDiscountCode": true
}
