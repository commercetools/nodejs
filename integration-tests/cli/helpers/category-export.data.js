export default [
  {
    key: 'fooKey',
    name: {
      en: 'categoryA',
    },
    slug: {
      en: 'category-a',
    },
    description: {
      en: 'descA',
    },
    externalId: 'a',
    assets: [],
  },
  {
    key: 'barKey',
    name: {
      en: 'categoryB',
    },
    slug: {
      en: 'category-b',
    },
    description: {
      en: 'descB',
    },
    externalId: 'b',
    assets: [],
  },
  {
    key: 'fooBarKey',
    name: {
      en: 'categoryC',
    },
    slug: {
      en: 'category-c',
    },
    description: {
      en: 'descC',
    },
    externalId: 'c',
    assets: [],
  },
]
