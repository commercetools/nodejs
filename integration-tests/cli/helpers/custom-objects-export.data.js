export default [
  {
    container: 'Ludus',
    key: 'copperKey',
    value: {
      paymentMethod: 'Cash',
      paymentID: '1',
    },
  },
  {
    container: 'Frobozz',
    key: 'jadeKey',
    value: {
      paymentMethod: 'cc',
      paymentID: '2',
    },
  },
  {
    container: 'Syrinx',
    key: 'crystalKey',
    value: {
      paymentMethod: 'new',
      paymentID: '3',
    },
  },
]
