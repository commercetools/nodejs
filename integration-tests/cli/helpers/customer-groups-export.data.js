export default [
  {
    groupName: 'Ludus',
    key: 'copperKey',
  },
  {
    groupName: 'Frobozz',
  },
  {
    groupName: 'Syrinx',
    key: 'crystalKey',
  },
]
