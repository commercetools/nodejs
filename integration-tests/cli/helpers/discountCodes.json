[{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "IT8MKKZX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlN85n0"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDHzsVX"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlb6Z1u"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITfWgUVd"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITpQgE1P"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITL2pvbZ"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITlrnkj5"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITDJO3LT"
},
{
  "name": {
    "en": "Sammy",
    "de": "Valerian"
  },
  "description": {
    "en": "greatest promo",
    "de": "super angebot"
  },
  "cartPredicate": "lineItemTotal(1 = 1) >  \"10.00 USD\"",
  "isActive": true,
  "maxApplications": 10,
  "maxApplicationsPerCustomer": 2,
  "code": "ITuIcDI2"
}]
