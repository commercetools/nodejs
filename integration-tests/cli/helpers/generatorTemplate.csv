name.en,name.de,description.en,description.de,cartPredicate,isActive,maxApplications,maxApplicationsPerCustomer
Sammy,Valerian,greatest promo,super angebot,some cart predicate,true,9,4
