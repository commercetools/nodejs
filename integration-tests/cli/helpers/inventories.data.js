export const inventories = [{
  sku: '12345',
  quantityOnStock: 20,
  custom: {
    type: {
      key: 'custom-type',
    },
    fields: {
      description: 'integration tests!! arrgggh',
    },
  },
}]

export const customFields = [{
  key: 'custom-type',
  name: {
    en: 'customized fields',
  },
  description: {
    en: 'custom description',
  },
  resourceTypeIds: ['inventory-entry'],
  fieldDefinitions: [{
    name: 'description',
    type: {
      name: 'String',
    },
    required: false,
    label: {
      en: 'owner',
    },
    inputHint: 'SingleLine',
  }],
}]
