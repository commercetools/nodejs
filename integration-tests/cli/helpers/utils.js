import { createAuthMiddlewareForClientCredentialsFlow } from '@commercetools/sdk-middleware-auth'
import { createClient } from '@commercetools/sdk-client'
import { createRequestBuilder } from '@commercetools/api-request-builder'
import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
import fetch from 'node-fetch'

export function clearData(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }
  return client.process(request, payload => {
    // Built-in states cannot be deleted
    const results =
      entityName === 'states'
        ? payload.body.results.filter(state => state.builtIn === false)
        : payload.body.results
    return Promise.all(
      results.map(result =>
        client.execute({
          uri: service
            .byId(result.id)
            .withVersion(result.version)
            .build(),
          method: 'DELETE',
        })
      )
    )
  })
}

export function createData(apiConfig, entityName, data, id) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })
  const requestOption = { projectKey: apiConfig.projectKey }
  const service = createRequestBuilder(requestOption)[entityName]
  return Promise.all(
    data.map(_data => {
      if (id) service.byId(id)
      const request = {
        uri: service.build(),
        method: 'POST',
        body: _data,
      }
      return client.execute(request)
    })
  )
}

export function getId(apiConfig, entityName) {
  const client = createClient({
    middlewares: [
      createAuthMiddlewareForClientCredentialsFlow({ ...apiConfig, fetch }),
      createHttpMiddleware({ host: apiConfig.apiUrl, fetch }),
    ],
  })

  const service = createRequestBuilder({
    projectKey: apiConfig.projectKey,
  })[entityName]

  const request = {
    uri: service.build(),
    method: 'GET',
  }

  return client
    .execute(request)
    .then(result => Promise.resolve(result.body.results[0].id))
}
