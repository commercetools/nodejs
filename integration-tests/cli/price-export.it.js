import fs from 'mz/fs'
import tmp from 'tmp'
import path from 'path'
import csv from 'csvtojson'
import { getCredentials } from '@commercetools/get-credentials'
import { exec } from 'mz/child_process'
import { version } from '@commercetools/price-exporter/package.json'
import {
  sampleProductType,
  sampleCustomerGroup,
  sampleChannel,
  sampleCustomType,
  createProducts,
} from './helpers/prices.data'
import { createData, clearData } from './helpers/utils'

let projectKey
if (process.env.CI === 'true') projectKey = 'price-export-integration-test'
else projectKey = process.env.npm_config_projectkey

describe('Price Exporter', () => {
  let apiConfig
  const productType = { typeId: 'product-type' }
  const customerGroup = { typeId: 'customer-group' }
  const channel = { typeId: 'channel' }
  const customType = { typeId: 'type' }
  const bin = './integration-tests/node_modules/.bin/price-exporter'

  beforeAll(async () => {
    const credentials = await getCredentials(projectKey)
    apiConfig = {
      host: 'https://auth.sphere.io',
      apiUrl: 'https://api.sphere.io',
      projectKey,
      credentials,
    }
    await clearData(apiConfig, 'products')

    await Promise.all([
      clearData(apiConfig, 'types'),
      clearData(apiConfig, 'channels'),
      clearData(apiConfig, 'productTypes'),
      clearData(apiConfig, 'customerGroups'),
    ])

    const [
      createdType,
      createdChannel,
      createdProductType,
      createdCustomerGroup,
    ] = await Promise.all([
      createData(apiConfig, 'types', [sampleCustomType]),
      createData(apiConfig, 'channels', [sampleChannel]),
      createData(apiConfig, 'productTypes', [sampleProductType]),
      createData(apiConfig, 'customerGroups', [sampleCustomerGroup]),
    ])
    customType.id = createdType[0].body.id
    channel.id = createdChannel[0].body.id
    productType.id = createdProductType[0].body.id
    customerGroup.id = createdCustomerGroup[0].body.id

    const sampleProducts = await createProducts(
      productType,
      customerGroup,
      channel,
      customType
    )
    await createData(apiConfig, 'products', sampleProducts)
  }, 10000)

  afterAll(async () => {
    await clearData(apiConfig, 'products')
    await Promise.all([
      clearData(apiConfig, 'types'),
      clearData(apiConfig, 'channels'),
      clearData(apiConfig, 'productTypes'),
      clearData(apiConfig, 'customerGroups'),
    ])
  })

  describe('CLI basic functionality', () => {
    it('should print usage information given the help flag', async () => {
      const [stdout, stderr] = await exec(`${bin} --help`)
      expect(stderr).toBeFalsy()
      expect(stdout).toMatchSnapshot()
    })

    it('should print the module version given the version flag', async () => {
      const [stdout, stderr] = await exec(`${bin} --version`)
      expect(stderr).toBeFalsy()
      expect(stdout).toBe(`${version}\n`)
    })
  })

  describe('JSON export', () => {
    it('should not resolve refs and export prices to file as json', async () => {
      const jsonFilePath = tmp.fileSync().name

      const [stdout, stderr] = await exec(
        `${bin} -o ${jsonFilePath} -p ${projectKey} --staged`
      )
      expect(stderr).toBeFalsy()
      expect(stdout).toMatchSnapshot()

      const data = await fs.readFile(jsonFilePath, { encoding: 'utf8' })

      const actual = JSON.parse(data)
      expect(actual.prices).toBeInstanceOf(Array)
      expect(actual.prices).toHaveLength(4)

      const expectedPrice = {
        value: {
          centAmount: 2500,
        },
        customerGroup: {
          id: 'customerGroupName',
        },
        channel: {
          id: 'my-channel-key',
        },
        validFrom: '2017-11-01T08:01:19.000Z',
        validUntil: '2019-11-01T08:01:21.000Z',
      }
      actual.prices.forEach(variantPrice => {
        expect(variantPrice['variant-sku']).toMatch(/variant-sku/)
        expect(variantPrice.prices[0]).toMatchObject(expectedPrice)
      })
    })
  })

  describe('CSV export', () => {
    it('should resolve refs and export only fields in template', async () => {
      const template = path.join(
        __dirname,
        'expected-output',
        'price-exporter-template.csv'
      )
      const csvFilePath = tmp.fileSync().name
      const [stdout, stderr] = await exec(
        `${bin} -o ${csvFilePath} -p ${projectKey} -f csv -s -i ${template}`
      )
      expect(stderr).toBeFalsy()
      expect(stdout).toMatchSnapshot()
      const data = await fs.readFile(csvFilePath, { encoding: 'utf8' })

      const expected = [
        'customType',
        'variant-sku',
        'customerGroup.groupName',
        'channel.key',
        'customField.loremIpsum',
        'value.currencyCode',
      ]
      csv({ flatKeys: true })
        .fromString(data)
        .on('json', jsonObj => {
          expect(Object.keys(jsonObj)).toEqual(expected)
          expect(jsonObj['customerGroup.groupName']).toBe('customerGroupName')
          expect(jsonObj['channel.key']).toBe('my-channel-key')
          expect(jsonObj.customType).toBe('my-custom-type-key')
        })
    })
  })
})
