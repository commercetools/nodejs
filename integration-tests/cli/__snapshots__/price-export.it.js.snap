// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Price Exporter CLI basic functionality should print usage information given the help flag 1`] = `
"
Usage: price-exporter [options]
Exports prices from the commercetools platform

Options:
      --help          Show help                                        [boolean]
      --version       Show version number                              [boolean]
  -i, --input         Path to CSV template.
  -o, --output        Path to output file.                   [default: \\"stdout\\"]
      --apiUrl        The host URL of the HTTP API service.
                     [default: \\"https://api.europe-west1.gcp.commercetools.com\\"]
      --authUrl       The host URL of the OAuth API service.
                    [default: \\"https://auth.europe-west1.gcp.commercetools.com\\"]
  -p, --projectKey    API project key.                                [required]
      --accessToken   CTP client access token.
                      Required scopes: ['view_products', 'view_customers',
                      'view_types']
  -d, --delimiter     Used CSV delimiter for template and output. [default: \\",\\"]
  -w, --where         Where predicate for products from which to fetch prices.
  -f, --exportFormat  Format for export.
                                      [choices: \\"csv\\", \\"json\\"] [default: \\"json\\"]
  -s, --staged        Specify if prices should be from all or published
                      products.                                        [boolean]
      --logLevel      Logging level: error, warn, info or verbose.
                                                               [default: \\"info\\"]
      --logFile       Path to file where to save logs.
                                                 [default: \\"price-exporter.log\\"]
"
`;

exports[`Price Exporter CSV export should resolve refs and export only fields in template 1`] = `
"info Starting Export
info Export operation completed successfully
"
`;

exports[`Price Exporter JSON export should not resolve refs and export prices to file as json 1`] = `
"info Starting Export
info Export operation completed successfully
"
`;
